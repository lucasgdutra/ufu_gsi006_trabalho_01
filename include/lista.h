/* --------------------------------------------------------------------------------------
   TAD Lista_Ordenada
   Também usaremos como estrutura de dados para nos dar suporte no projeto, as Listas, e
   usaremos Listas Encadeadas e Ordenadas.
   Objetivo: Trabalho de disciplina Estrutura de dados (GSI006) BSI - Facom -
UFU Restricoes:

   Autor: Dalmo Scalon Inacio/ Lucas Gabriel Dutra de Souza
   Disciplina: Estrutura de Dados 1
   Data da ultima atualizacao: 04/01/2023
----------------------------------------------------------------------------------------*/
#ifndef LISTA_H
#define LISTA_H

/*
	Struct Node_Lista, que contem um inteiro indicando posicao, os dados que é um ponteiro
	do tipo void, para comportar qualquer tipo de dado e teremos mais um node indicando o
	proximo_node.

*/
typedef struct node_lista {
	int pos;
	void *dados;
	struct node_lista *proximo_node;
} Node_Lista;

/*
	Struct Lista, contem um Node_Lista indicando a primeira posicao da lista, int indicando
	o tamanho da lista, outro int pra receber o tamanho da estrutura.
*/
typedef struct lista {
	Node_Lista *primeiro;
	int tamanho_lista;
	int tamanho_estrutura_dados;
} Lista_Ordenada;

/*
 Nome: inicializa_lista
 Entrada: Ponteiro do tipo Lista e um int indicando o tamanho da estrutura.
 Processo: Verifica validade do ponteiro, e inicializa o primeiro com NULL,
 tamanho da lista com 0, e o tamanho da estrutura.
 Saida: return 0 caso dê certo, e 1 caso dê errado.
 */
int inicializa_lista(Lista_Ordenada *lista, int tamanho_estrutura_dados);

/*
 Nome: adiciona_em_lista
 Entrada: Ponteiro do tipo Lista, ponteiro do tipo void passando os dados, e um int indicando
 a posicao
 Processo: Verifica validade dos ponteiros, e adiciona elemento no final da lista
 Saida: return 0 caso dê certo, e 1 caso dê errado.
*/
int adiciona_em_lista(Lista_Ordenada *lista, void *dados, int pos);

/*
 Nome: remove_da_lista
 Entrada: Ponteiro do tipo Lista, ponteiro do tipo void passando o valor retirado
 e um int indicando a posicao
 Processo: Verifica validade dos ponteiros, e remove elemento do inicio da lista e passa
 o elemento removido para *valor.
 Saida: return 0 caso dê certo, e 1 caso dê errado.
*/
int remove_da_lista(Lista_Ordenada *lista, void *valor, int pos);

/*
 Nome: limpa_lista
 Entrada: Ponteiro do tipo Lista.
 Processo: Verifica validade do ponteiro, e remove todos elementos da lista passada.
 Saida: return 0 caso dê certo, e 1 caso dê errado.
*/
int limpa_lista(Lista_Ordenada *lista);


/*
 Nome: retorna_tamanho_lista
 Entrada: Ponteiro do tipo Lista, ponteiro de inteiro indicando o tamanho da lista
 Processo: Verifica validade dos ponteiros, e passa o tamanho da lista para a variavel
 recebida na função.
 Saida: return 0 caso dê certo, e 1 caso dê errado.
 */
int retorna_tamanho_lista(Lista_Ordenada *lista, int *tam);

/*
 Nome: buscar_item_lista
 Entrada: Ponteiro do tipo Lista, ponteiro de inteiro indicando a posicao, e ponteiro do tipo
 void indicando os dados que serão retornados.
 Processo: Verifica validade dos ponteiros, procura na lista comparando por posicao,
 se encontrado posiçoes iguais, retornará os dados.
 Saida: return 0 caso dê certo, e 1 caso dê errado.
 */
int buscar_item_lista(Lista_Ordenada *lista, int posicao, void *retorno_dados);

#endif
