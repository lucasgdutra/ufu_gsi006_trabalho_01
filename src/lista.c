#include "../include/lista.h"
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

int inicializa_lista(Lista_Ordenada *lista, int tamanho_estrutura_dados) {
	// testa se ponteiro eh valido
	if (lista == NULL) {
		return 1;
	}
	lista->primeiro = NULL;
	lista->tamanho_estrutura_dados = tamanho_estrutura_dados;
	lista->tamanho_lista = 0;
	return 0;
}

int adiciona_em_lista(Lista_Ordenada *lista, void *dados, int pos) {
	// testa se ponteiro eh valido
	if (lista == NULL || dados == NULL) {
		return 1;
	}
	// instancia estrutura para receber elemento
	Node_Lista *novo_node;
	novo_node = (Node_Lista *)malloc(sizeof(Node_Lista));
	if (novo_node == NULL) {
		// erro ao alocar elemento
		return 1;
	}
	novo_node->dados = malloc(lista->tamanho_estrutura_dados);
	if (novo_node->dados == NULL) {
		// erro ao alocar dados do elemento
		free(novo_node);
		return 1;
	}

	// define propriedades do nodo
	memcpy(novo_node->dados, dados, lista->tamanho_estrutura_dados);
	novo_node->pos = pos;
	novo_node->proximo_node = NULL;

	// insere node na lista
	if (lista->primeiro == NULL) {
		// se lista vazia insere na primeira posição
		lista->primeiro = novo_node;
	} else {
		Node_Lista *node_anterior = NULL;
		Node_Lista *node_atual = lista->primeiro;

		while (true) {

			if (node_atual == NULL) {

				// inserir na ultima posicao
				node_anterior->proximo_node = novo_node;
				break;
			} else if (node_anterior == NULL) {
				// inserir na primeira posição
				if (pos < node_atual->pos) {
					lista->primeiro = novo_node;
					novo_node->proximo_node = node_atual;
					break;
				}
			} else if (pos >= node_anterior->pos && pos < node_atual->pos) {
				// inserir no meio
				node_anterior->proximo_node = novo_node;

				novo_node->proximo_node = node_atual;
				break;
			}

			node_anterior = node_atual;
			node_atual = node_atual->proximo_node;
		}
	}

	// ajusta tamanho da lista
	lista->tamanho_lista++;
	return 0;
};

int remove_da_lista(Lista_Ordenada *lista, void *valor, int pos) {
	// testa se ponteiro eh valido
	if (lista == NULL) {
		return 1;
	}
	if (lista->tamanho_lista > 0) {
		Node_Lista *node_anterior = NULL;
		Node_Lista *node_atual = lista->primeiro;
		while (true) {
			if (node_atual == NULL) {
				return 1;
			}
			if (node_atual->pos == pos) {
				// guarda endereço do node que vai ser removido
				Node_Lista *node_removido = node_atual;

				// copia dados do node removido para ponteiro valor em argumento
				// da função, se valor for NULL continua funcao sem copiar dados
				if (valor != NULL) {
					memcpy(valor, node_removido->dados,
						   lista->tamanho_estrutura_dados);
				}
				// ajusta apontamento do node anterior para proximo node
				if (node_anterior == NULL) {
					// removendo primeiro elemento
					lista->primeiro = node_atual->proximo_node;
				} else {
					node_anterior->proximo_node = node_atual->proximo_node;
				}

				// libera memoria de node_removido
				free(node_removido->dados);
				free(node_removido);

				// ajusta tamanho da lista
				lista->tamanho_lista--;
				break;
			}
			node_anterior = node_atual;
			node_atual = node_atual->proximo_node;
		}

		return 0;
	} else {
		return 1;
	}
}

int limpa_lista(Lista_Ordenada *lista) {
	// testa se ponteiro eh valido
	if (lista == NULL) {
		return 1;
	}
	while (lista->tamanho_lista > 0) {
		Node_Lista *node_removido = lista->primeiro;
		lista->primeiro = node_removido->proximo_node;
		// libera memoria de node_removido
		free(node_removido->dados);
		free(node_removido);
		lista->tamanho_lista--;
	}
	return 0;
}

int retorna_tamanho_lista(Lista_Ordenada *lista, int *tam) {
	// testa se ponteiro eh valido
	if (lista == NULL || tam == NULL) {
		return 1;
	}
	*tam = lista->tamanho_lista;
	return 0;
}

int buscar_item_lista(Lista_Ordenada *lista, int posicao, void *retorno_dados) {
	// testa se ponteiro eh valido
	if (lista == NULL || retorno_dados == NULL) {
		return 1;
	}
	Node_Lista *node_atual = lista->primeiro;
	if (node_atual == NULL) {
		// lista vazia
		return 1;
	}
	while (true) {
		if (node_atual == NULL) {
			// node_atual chegou ao final da lista, item nao encontrado
			return 1;
		}
		if (node_atual->pos == posicao) {
			memcpy(retorno_dados, node_atual->dados,
				   lista->tamanho_estrutura_dados);
			return 0;
		}
		if (posicao < node_atual->pos) {
			// como a lista eh ordenada, se a posicao da busca for menor que a
			// posicao atual entao a posicao da busca já passou e nao existe
			return 1;
		}
		node_atual = node_atual->proximo_node;
	}
	return 0;
}