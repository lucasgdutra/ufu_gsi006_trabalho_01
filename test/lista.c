#include "../include/lista.h"
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>

int main() {
    Lista_Ordenada l;
    inicializa_lista(&l, sizeof(int));

    int entrada;
    while (scanf("%d", &entrada) != EOF) {
        int s = adiciona_em_lista(&l, &entrada, entrada);
        if (s == 1)
            return 1;
    }

    while (l.tamanho_lista > 0) {

        if (l.primeiro != NULL) {
            Node_Lista *node_removido = l.primeiro;
            l.primeiro = node_removido->proximo_node;
            printf("dados:%d, pos:%d, tam_lista:%d\n",
                   *((int *)node_removido->dados), node_removido->pos,
                   l.tamanho_lista);
            free(node_removido->dados);
            free(node_removido);
            l.tamanho_lista--;
        } else {
            return 1;
        }
    }

    return 0;
}