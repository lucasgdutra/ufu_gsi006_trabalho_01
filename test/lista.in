1
8
6
9
2
10
0
-1