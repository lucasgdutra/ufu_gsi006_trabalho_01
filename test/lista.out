dados:-1, pos:-1, tam_lista:8
dados:0, pos:0, tam_lista:7
dados:1, pos:1, tam_lista:6
dados:2, pos:2, tam_lista:5
dados:6, pos:6, tam_lista:4
dados:8, pos:8, tam_lista:3
dados:9, pos:9, tam_lista:2
dados:10, pos:10, tam_lista:1
